buffer/src/main/java/io/netty/buffer/TestingPooledByteBufAllocator.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.buffer;
+
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakTracker;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A {@link ByteBufAllocator} for unit testing.  This allocator performs leak tracking
+ * on all allocated buffers.  At the end the test, call {@link #close} to assert
+ * that all buffers have been freed.
+ */
+public class TestingPooledByteBufAllocator extends PooledByteBufAllocator implements Closeable {
+
+    private final TestingResourceLeakDetector leakDetector = new TestingResourceLeakDetector();
+
+    public TestingPooledByteBufAllocator() {
+    }
+
+    public TestingPooledByteBufAllocator(boolean preferDirect) {
+        super(preferDirect);
+    }
+
+    @Override
+    ByteBuf toLeakAwareBufferInternal(ByteBuf buf) {
+        ResourceLeakTracker<ByteBuf> leak = leakDetector.track(buf);
+        if (leak != null) {
+            buf = new AdvancedLeakAwareByteBuf(buf, leak);
+        }
+        return buf;
+    }
+
+    @Override
+    CompositeByteBuf toLeakAwareBufferInternal(CompositeByteBuf buf) {
+        ResourceLeakTracker<ByteBuf> leak = leakDetector.track(buf);
+        if (leak != null) {
+            buf = new AdvancedLeakAwareCompositeByteBuf(buf, leak);
+        }
+        return buf;
+    }
+
+    /**
+     * Throws an exception if all allocated buffers have been not freed.
+     * @throws AssertionError if buffers have been leaked.
+     */
+    @Override
+    public void close() {
+        if (leakDetector.hadLeak()) {
+            throw new AssertionError("LEAK");
+        }
+    }
+
+    private static class TestingResourceLeakDetector extends ResourceLeakDetector<ByteBuf> {
+
+        private final AtomicBoolean leaked = new AtomicBoolean();
+
+        public TestingResourceLeakDetector() {
+            super(ByteBuf.class);
+        }
+
+        @Override
+        protected boolean shouldTrack(ByteBuf obj) {
+            return true;
+        }
+
+        @Override
+        protected boolean isLeakReportingEnabled() {
+            return true;
+        }
+
+        @Override
+        protected void reportTracedLeak(String resourceType, String records) {
+            leaked.set(true);
```

Comment:
I wonder if we should better store the resourceType and records here as well to make it easier to debug ?

We could include these in the AssertationError.

Reply:
I can do that.  Do you want a new public exception class so you can pull out the records programmatically, or private one so you need a debugger?  I use a debugger when working on stuff like this, so I don't have a strong opinion.